Extract tunnel options to simplify override, and inject them through hive

giorio94 opened the pull request:
Certain features (e.g., egress gateway, high-scale ipcache, ...) require
the configuration of the tunnel device also then the primary routing
mode is set to native routing, as some traffic may need to flow through
it. Currently, every affected code path needs ad-hoc checks, which is
getting more and more complex with the addition of new features.

This commit consolidates the tunnel-related options, that is protocol,
port, device name and whether it requires MTU adaptation, into a single
struct, initialized based on the user configuration and the requests of
additional features, and propagated through hive. Specifically, different
modules can now request the configuration of the tunnel device (regardless
of the primary routing mode) through the injection of the appropriate
enabler object. Additional validation constraints can be specified if a
given feature is compatible only with a specific encapsulation protocol.

This change introduces a small functional difference, because now the
tunnel protocol flag is always respected, and no longer defaulted to
geneve when Cilium is configured in native routing mode, and DSR
Geneve is enabled. This prevents causing unexpected tunnel changes
when turning on/off other features (DSR Geneve in this case).

The primary routing mode is still represented by the TunnelingEnabled()
function, both to reduce the scope of this change and because it is
currently accessed also before the initialization of the hive.

Fixes: https://github.com/cilium/cilium/issues/25769
~~Depends on: https://github.com/cilium/cilium/pull/29053~~

giorio94 committed:
tunnel: extract options in dedicated object and inject through hive

Certain features (e.g., egress gateway, high-scale ipcache, ...) require
the configuration of the tunnel device also then the primary routing
mode is set to native routing, as some traffic may need to flow through
it. Currently, every affected code path needs ad-hoc checks, which is
getting more and more complex with the addition of new features.

This commit consolidates the tunnel-related options, that is protocol,
port, device name and whether it requires MTU adaptation, into a single
struct, initialized based on the user configuration and the requests of
additional features, and propagated through hive. Specifically, different
modules can now request the configuration of the tunnel device (regardless
of the primary routing mode) through the injection of the appropriate
enabler object. Additional validation constraints can be specified if a
given feature is compatible only with a specific encapsulation protocol.

This change introduces a small functional difference, because now the
tunnel protocol flag is always respected, and no longer defaulted to
geneve when Cilium is configured in native routing mode, and DSR
Geneve is enabled. This prevents causing unexpected tunnel changes
when turning on/off other features (DSR Geneve in this case).

The primary routing mode is still represented by the TunnelingEnabled()
function, both to reduce the scope of this change and because it is
currently accessed also before the initialization of the hive.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

christarazi reviewed (approved):
ci-structure LGTM

jschwinger233 commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -42,7 +43,7 @@ import (
 // if this function cannot determine the strictness an error is returned and the boolean
 // is false. If an error is returned the boolean is of no meaning.
-func initKubeProxyReplacementOptions() error {
+func initKubeProxyReplacementOptions(tunnelConfig tunnel.Config) error {
```
I see reasons to pass tunnelConfig as a parameter, and I understand there already have been a `mtuConfig` under `Daemon` struct so it makes sense to put `tunnelConfig` in the same place. Still, this sets a precedent to pass config everywhere instead of using the global `option.Config`. It's not this PR's fault, and I don't have a easy solution either. Maybe we can consider moving these configs into option pkg in future, while it goes beyond this PR's scope.

giorio94 replied:
My understanding is that the long term goal is to get rid as much as possible of the global `option.Config` object, and move the various settings close to the corresponding cell. This is what I've been seeing as part of the refactoring of various modules to the hive/cell paradigm, e.g., \[1\]. IMO, passing config objects as parameter has the advantage of making their usage more explicit, rather than just accessing a global variable/function deep in the code path (although at the cost of having to propagate it), and simplify at the same time unit testing, as we depend less on global state.

Now, I agree that putting `tunnelConfig` under `Daemon` is not great. Ideally, consumers (like the KPR module in this case) should also be cells, and could just depend on the `tunnelConfig` itself. The current approach is mostly a tradeoff to limit the scope of the single refactoring. Long term, these variables should be removed from `Daemon`.

\[1\]: https://github.com/cilium/cilium/pull/28619#discussion_r1381395461

jschwinger233 reviewed (approved):
LGTM, thanks! Left a comment but don't let it block merging.

lmb reviewed (approved):
Loader changes LGTM.

julianwiedmann commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -213,23 +214,22 @@ func initKubeProxyReplacementOptions() error {
-			return fmt.Errorf("Node Port %q mode with %s dispatch cannot be used with %s tunneling.",
-				option.Config.NodePortMode, option.Config.LoadBalancerDSRDispatch, option.Config.TunnelProtocol)
+			return fmt.Errorf("Tunnel routing with Node Port %q mode requires %s dispatch.",
+				option.Config.NodePortMode, option.Config.LoadBalancerDSRDispatch)
```
```suggestion
				option.Config.NodePortMode, option.DSRDispatchGeneve)
```

julianwiedmann commented on test/k8s/services.go:
```diff
@@ -584,6 +586,7 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`,
 				"loadBalancer.mode":         "hybrid",
 				"loadBalancer.algorithm":    "random",
 				"routingMode":               "native",
+				"tunnelProtocol":            "geneve",
```
Does that mean we're no longer auto-selecting `tunnelProtocol: geneve` when using DSR-GENEVE in native-routing? If it's possible without too much hassle, I think it would be good to preserve existing behaviour for now (to avoid upgrade regressions).

giorio94 replied:
> Does that mean we're no longer auto-selecting tunnelProtocol: geneve when using DSR-GENEVE in native-routing?

Yes. I'd personally prefer to keep this explicit, to avoid mutating the tunnel protocol configured by the users unexpectedly. Allowing the protocol to be overridden by a given module would also open up for additional complexity and possible conflicts (as different features may request different protocols) to be resolved.

But I agree that this is a possible upgrade regression. My idea was to call it out in the upgrades notes (I haven't pushed it yet though). Do you think that would be sufficient? (If users overlook that, then the agents will fail at startup with an error mentioning the protocol incompatibility). Another possibility could be adding such defaulting at the helm level, maybe setting the protocol to geneve if left to the default value, and DSR-GENEVE is enabled.

julianwiedmann replied:
> But I agree that this is a possible upgrade regression. My idea was to call it out in the upgrades notes (I haven't pushed it yet though).

Realistically yes, that's ok. There shouldn't be many users around, so let's not cause too much extra work for a corner-case. (adding `upgrade-impact` just in case)

There is a more general point here - imho Cilium should auto-select such config options whenever possible, and only complain on conflicts. Maybe it's ok to just have this on a ~~high~~ helm level. But we won't solve that today :slightly_smiling_face:.

julianwiedmann reviewed (approved):
One log message nitpick, and one question wrt to upgrade behaviour. Both could be addressed as follow-on PR if needed. Otherwise lgtm, thank you very much Marco!

ysksuzuki commented on daemon/cmd/daemon.go:
```diff
@@ -218,6 +219,9 @@ type Daemon struct {
 	healthProvider cell.Health
+
+	// Tunnel-related configuration
+	tunnelConfig tunnel.Config
```
Sorry for my newbie question.

The DI container instantiates a tunnel.Config and injects it here automatically so that features relying on ad-hoc tunnel such as DSR GENEVE and Egress GW can retrieve the tunnel config as Single Source of Truth? Is that this PR aiming at?

giorio94 replied:
Yes, essentially it gets injected into any provide and invoke function that declares it as parameter (either directly or through a struct with the cell.In field). More specifically, in this specific case it gets injected [here](https://github.com/cilium/cilium/blob/7a6ed1452e404f2e9af75d0d4dae552b802e357c/daemon/cmd/daemon_main.go#L1681), and then just copied into the Daemon struct. This allows to have a single source of truth about whether e.g., the tunnel device needs to be created and its configuration, without depending on any global variable/function.

Leveraging the same DI strategy, this PR also allows additional modules (e.g., egress gateway/DSR-geneve) to specifically request that the tunnel device (and alike) are configured even if Cilium is configured in native routing mode. But it also guarantees that any consumer of `tunnel.Config` will always see the final materialized configuration (thanks to the DAG constructed by hive/dig for the initialization order).

ysksuzuki reviewed (approved):
LGTM! Just a non-blocking newbie question.

derailed commented on pkg/datapath/loader/netlink.go:
```diff
@@ -547,7 +548,7 @@ func setupGeneveDevice(dport, mtu int) error {
 // Changing the port will recreate the device. Changing the MTU will modify the
 // device without recreating it.
-func setupVxlanDevice(port, mtu int) error {
+func setupVxlanDevice(port uint16, mtu int) error {
```
Wondering why we change the sig here from int->uint since that func ends up promoting to int?
Also the const are defined as int?

giorio94 replied:
It just seemed to me that it was better to propagate the correct type (given that this represents a port) as deep as possible in the call stack. Unfortunately, the library field is an int, so we need the conversion (differently from the geneve case, which correctly accepts it as uint16). I've also updated the constants to explicitly specify the type.

derailed commented on pkg/datapath/loader/netlink.go:
```diff
@@ -506,7 +507,7 @@ func setupTunnelDevice(mode string, port, mtu int) error {
 // Changing the destination port will recreate the device. Changing the MTU will
 // modify the device without recreating it.
-func setupGeneveDevice(dport, mtu int) error {
+func setupGeneveDevice(dport uint16, mtu int) error {
```
Should we dry this up? ie setupDevice(mode, port, mtu) or use generics?? likely a separate PR...

giorio94 replied:
Yeah, I'd leave that out of this PR to avoid extending too much its scope.

derailed commented on pkg/datapath/tunnel/cell.go:
```diff
@@ -0,0 +1,51 @@
+	"github.com/cilium/cilium/pkg/option"
+)
+
+var Cell = cell.Module(
```
nit: Docs What does this do?

derailed commented on pkg/datapath/tunnel/tunnel.go:
```diff
@@ -0,0 +1,220 @@
+	TunnelPort     uint16
+}
+
+func (def userCfg) Flags(flags *pflag.FlagSet) {
```
nit: Docs?

derailed reviewed (commented):
@giorio94 Nice work Marco!

giorio94 committed:
docs: add upgrade note about tunnel protocol with DSR Geneve

Historically, the tunnel protocol was defaulted to geneve when Cilium
was configured in native routing mode and DSR Geneve was enabled. This
is no longer the case since the last commit, and the tunnel protocol
flag is now always respected (possibly triggering an error if
incompatible), so that we don't cause unexpected tunnel changes when
turning on/off other features. Hence, let's document in the upgrade
notes that users now need to explicitly configure the tunnel protocol
to geneve when using DSR Geneve.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

derailed reviewed (approved):
@giorio94 Thanks for the updates!
